cli/internal/taskhash/taskhash.go

```diff
@@ -242,28 +244,74 @@ func (th *Tracker) CalculateTaskHash(logger hclog.Logger, packageTask *nodes.Pac
 		return "", fmt.Errorf("cannot find package-file hash for %v", packageTask.TaskID)
 	}
 
-	var keyMatchers []string
-	var framework *inference.Framework
-	envVarContainingExcludePrefix := ""
+	allEnvVarMap := env.EnvironmentVariableMap{}
+	explicitEnvVarMap := env.EnvironmentVariableMap{}
+	matchingEnvVarMap := env.EnvironmentVariableMap{}
 
+	var framework *inference.Framework
 	if frameworkInference {
-		envVarContainingExcludePrefix = "TURBO_CI_VENDOR_ENV_KEY"
+		// See if we infer a framework.
 		framework = inference.InferFramework(packageTask.Pkg)
-		if framework != nil && framework.EnvMatcher != "" {
-			// log auto detected framework and env prefix
-			logger.Debug(fmt.Sprintf("auto detected framework for %s", packageTask.PackageName), "framework", framework.Slug, "env_prefix", framework.EnvMatcher)
-			keyMatchers = append(keyMatchers, framework.EnvMatcher)
+		if framework != nil {
+			logger.Debug(fmt.Sprintf("auto detected framework for %s", packageTask.PackageName), "framework", framework.Slug, "env_prefix", framework.EnvWildcards)
+
+			computedWildcards := []string{}
+			computedWildcards = append(computedWildcards, framework.EnvWildcards...)
+
+			// Vendor excludes are only applied against inferred includes.
+			excludePrefix, exists := th.EnvAtExecutionStart["TURBO_CI_VENDOR_ENV_KEY"]
+			if exists && excludePrefix != "" {
+				computedExclude := "!" + excludePrefix + "*"
+				logger.Debug(fmt.Sprintf("excluding environment variables matching wildcard %s", computedExclude))
+				computedWildcards = append(computedWildcards, computedExclude)
+			}
+
+			inferenceEnvVarMap, err := th.EnvAtExecutionStart.FromWildcards(computedWildcards)
+			if err != nil {
+				return "", err
+			}
+
+			userEnvVarSet, err := th.EnvAtExecutionStart.FromWildcardsUnresolved(packageTask.TaskDefinition.EnvVarDependencies)
+			if err != nil {
+				return "", err
+			}
+
+			allEnvVarMap.Union(userEnvVarSet.Inclusions)
+			allEnvVarMap.Union(inferenceEnvVarMap)
+			allEnvVarMap.Difference(userEnvVarSet.Exclusions)
+
+			explicitEnvVarMap.Union(userEnvVarSet.Inclusions)
+			explicitEnvVarMap.Difference(userEnvVarSet.Exclusions)
+
+			matchingEnvVarMap.Union(inferenceEnvVarMap)
+			matchingEnvVarMap.Difference(userEnvVarSet.Exclusions)
```

Comment:
I considered it, I really did. The part that is missing from my calculation here was that, when originally authored, they _were_ different because the global had to support `*THASH*`. Also, I want to figure out a way to _drop_ `_defaultEnvVars` which would reduce the global one to just `FromWildcards`.

There was even a point where I considered leaving `GetHashableEnvVars` implemented via `FromWildcards` in the codebase, which _was_ that abstraction layer:

https://github.com/vercel/turbo/pull/5017/commits/bba955fdf683f74321345737b07d545eeb6a0740#diff-79ab7e6ee57b9aa25276151ceb78d11d64a8496d30979203412789cfd8970eb5R277

I ultimately decided against that too. I decided that I already had evidence that these two things changed independently and that their current ~identical implementation is just (currently) convergent evolution.